draft/XXXX-configuration_handling/README.md

```diff
@@ -0,0 +1,202 @@
+certain kind of "additive" effect as well. The idea here is to vary a set of
+parameters as a group. For example:
+
+```yaml
+a: 0
+b:
+  lockstep: foo
+  value: [1, 2, 3]
+c:
+  lockstep: foo
+  value: [4, 5, 6]
+```
+
+yielding the following `(a, b, c)` ordered triples:
+
+```
+(0, 1, 4)
+(0, 2, 5)
+(0, 3, 6)
+```
+
+Note that this requires specifying which "lockstep group" each parameter is a
+part of (identified by a string, in this case `foo`). Note also that `a = 0`
+once more, as it is specified as a plain value.
+
+#### Ranges
+
+We may want to support different types of range operators. The `value` field in
+the foregoing examples takes a YAML list as its argument, explicitly listing the
+values to take on. However, something like the following:
+
+```yaml
+a: 0
+b:
+  value:
+    sequence:
+      start: 1
+      stop: 3
+      step: 1
+c:
+  value:
+    sequence:
+      start: 4
+      stop: 6
+```
+
+would be the equivalent of the combinatorial example above (with the `step`
+parameter defaulting to `1`, as in the case of `c`).
+
+Other ranges can be implemented as needed.
+
+#### SMQTK algorithms
+
+Tinker Engine includes first-class support for SMQTK. A special
+meta-configuration option will enable SMQTK algorithms to be used as swappable
+values during configuration processing. For example:
+
+```yaml
+foo:
+  value:
+    - smqtk: MyAlgorithm.FooBar
+      config:
+        foo: 1
+        bar: 2
+    - smqtk: MyAlgorithm.FooBaz
+      config:
+        foo: 1
+        bar: 2
+```
+
+This will cause the `foo` value to take on a SMQTK algorithm instance as
+directed by the arguments. The runtime will first check to see that
+`MyAlgorithm.FooBar` and `MyAlgorithm.FooBaz` are existing, usable SMQTK
+algorithms as part of the validation phase.
+
+#### Schema and validation
+
+The meta-configuration schema will be formalized in such a way that the format
+can be presented in documentation easily, and so that configurations can be
+validated at runtime. A tool such as JSON Schema may be the right tool for this.
+
+#### Simple configuration
+
+As a special case, the meta-configuration can forgo any of these parameter
+varying options to produce a single configuration with all values statically
+set.
+
+### Programmatic Configuration
+
+As a fallback, the user can forgo using a declarative configuration at all and
+instead define a Python generator function that yields one or more simple
+configuration values itself. This could be useful if the declarative mechanisms
+are not expressive enough to achieve some result.
+
+As a matter of software architecture, the declarative meta-configuration logic
+will be implemented as a function of this type, that includes logic for parsing
+and interpreting the meta-configuration file, yielding concrete implementations
+as appropriate.
+
+## Reference Implementation
+
+Because this RFC is light on specific, concrete implementation direction, a
```

Comment:
Something missing from here are: 
- Serializing requirements for config?
- Variable interpolation (https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#variable-interpolation)
- Should we have structured configs with type safety?  
- Merging configs: (https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#merging-configurations)
- Config Groups: Referencing another config in your config and pulling that into the place it's reference.  This is useful for things like  if we have a few configurations for an algorithm, we don't need a separate global config of that.  Say if we have two different servers with very different configs, it would be cleaner to just pull it in.  e.g. https://hydra.cc/docs/tutorials/basic/your_first_app/config_groups
- jupyter notebook support?

Reply:
> Something missing from here are:
> 
> * Serializing requirements for config?

What do you mean by this?

> * Variable interpolation (https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#variable-interpolation)

It is totally plausible that we would just use OmegaConf as the config parsing infrastructure, and thereby gain all of these neat features that it offers.

> * Should we have structured configs with type safety?

At this stage it doesn't seem absolutely necessary (since we have nothing, and need to get to one quantum of functionality, which wouldn't include type safety) but if this is something OmegaConf offers then we could get this "for free" if we adopt OmegaConf.

> * Merging configs: (https://omegaconf.readthedocs.io/en/2.0_branch/usage.html#merging-configurations)

(Same comment about riding OmegaConf's coattails.)

> * Config Groups: Referencing another config in your config and pulling that into the place it's reference.  This is useful for things like  if we have a few configurations for an algorithm, we don't need a separate global config of that.  Say if we have two different servers with very different configs, it would be cleaner to just pull it in.  e.g. https://hydra.cc/docs/tutorials/basic/your_first_app/config_groups

YAML allows this by default (which is a big reason I'm advocating for YAML rather than straight JSON).

> * jupyter notebook support?

This seems like an entirely separate concern. Does it affect the config handling infrastructure?

Reply:
Serializing requirements for config:
-  I was thinking about the limitations.  What if set one of the parameters to being a pytorch tensor? 
- This leads into a bigger question, how do we add config options during run?  Say we want to add which network was chosen by our system.  This is related to saving evaluation.  

I agree, using Omega config makes sense here for variable interpolation and merging configs.  

Config Groups:  YAML is able to automatically pull in other YAML files when referenced and maintain the structure.  That's pretty cool

The jupyter notebook issue is to make sure these parameters can be on the run from the console, not only by running the framework cmd calling.

Reply:
> Serializing requirements for config:
> 
> * I was thinking about the limitations.  What if set one of the parameters to being a pytorch tensor?

Right now, you would supply the parameters needed to set up a pytorch tensor, then you'd use the values at runtime to actually construct the tensor object.

It is possible that we could add a config-time mechanism for constructing certain things automatically (which is precisely what we're proposing to add in this RFC for SMQTK algorithm objects).

> * This leads into a bigger question, how do we add config options during run?  Say we want to add which network was chosen by our system.  This is related to saving evaluation.

I don't think this would affect config. You'd make some decision at runtime about, e.g., which network was selected, then you would use that network and maybe record the choice in your outputs.

> The jupyter notebook issue is to make sure these parameters can be on the run from the console, not only by running the framework cmd calling.

Ok. There are already issues filed for this (https://github.com/tinker-engine/tinker-engine/issues/35, https://github.com/tinker-engine/tinker-engine/issues/102) so even if we don't expose the functionality in this PR, we have definite plans for doing so already.

Reply:
For the serialization, what about program outputs (non-metric) but are values we want to track in the exact same way as parameters.

Reply:
If these are outputs you want to track, then I'd really prefer if you open an issue to discuss this there. This PR is specifically focused on how to specify a meta-configuration, how to process it into concrete configurations, and then how to run the protocol once per concrete configuration. How to specify/track outputs is a completely separate topic.

Reply:
Let move this to another RFC after changing the name of this